sdk/core/core-rest-pipeline/test/formDataPolicy.spec.ts

```diff
@@ -0,0 +1,211 @@
+import { BlobLike, BodyPart, FormDataMap, MultipartRequestBody } from "../src/interfaces";
+import { isNode } from "@azure/core-util";
+import { isBlobLike, isNodeReadableStream } from "../src/util/stream";
+import { Readable } from "stream";
+
+describe("formDataPolicy", function () {
+  afterEach(function () {
+    sinon.restore();
+  });
+
+  it("prepares x-www-form-urlencoded form data correctly", async function () {
+    const request = createPipelineRequest({
+      url: "https://bing.com",
+      headers: createHttpHeaders({
+        "Content-Type": "application/x-www-form-urlencoded",
+      }),
+    });
+    request.formData = {
+      service: "registry.azurecr.io",
+      scope: "repository:library/hello-world:metadata_read",
+    };
+    const successResponse: PipelineResponse = {
+      headers: createHttpHeaders(),
+      request,
+      status: 200,
+    };
+    const next = sinon.stub<Parameters<SendRequest>, ReturnType<SendRequest>>();
+    next.resolves(successResponse);
+
+    const policy = formDataPolicy();
+
+    const result = await policy.sendRequest(request, next);
+
+    assert.isUndefined(result.request.formData);
+    assert.strictEqual(
+      result.request.body,
+      `service=registry.azurecr.io&scope=repository%3Alibrary%2Fhello-world%3Ametadata_read`
+    );
+  });
+
+  it("prepares x-www-form-urlencoded form data correctly for array value", async function () {
+    const request = createPipelineRequest({
+      url: "https://bing.com",
+      headers: createHttpHeaders({
+        "Content-Type": "application/x-www-form-urlencoded",
+      }),
+    });
+    request.formData = { a: "va", b: "vb", c: ["vc1", "vc2"] };
+    const successResponse: PipelineResponse = {
+      headers: createHttpHeaders(),
+      request,
+      status: 200,
+    };
+    const next = sinon.stub<Parameters<SendRequest>, ReturnType<SendRequest>>();
+    next.resolves(successResponse);
+
+    const policy = formDataPolicy();
+
+    const result = await policy.sendRequest(request, next);
+
+    assert.isUndefined(result.request.formData);
+    assert.strictEqual(result.request.body, `a=va&b=vb&c=vc1&c=vc2`);
+  });
+
+  describe("multipart/form-data", function () {
+    async function performRequest(formData: FormDataMap) {
+      const request = createPipelineRequest({
+        url: "https://bing.com",
+        headers: createHttpHeaders({
+          "Content-Type": "multipart/form-data",
+        }),
+        formData,
+      });
+      const successResponse: PipelineResponse = {
+        headers: createHttpHeaders(),
+        request,
+        status: 200,
+      };
+      const next = sinon.stub<Parameters<SendRequest>, ReturnType<SendRequest>>();
+      next.resolves(successResponse);
+
+      const policy = formDataPolicy();
+
+      return policy.sendRequest(request, next);
+    }
+
+    it("prepares a form with multiple fields correctly", async function () {
```

Comment:
can we add a test with fancy multibyte unicode characters?

Reply:
made this test case a bit spookier